File: frontend/coordinator-android/engine/bin/main/ug/qaat/engine/CheckinValidator.kt
package ug.qaat.engine

import ug.qaat.crypto.QrVerify
import ug.qaat.crypto.VaultCrypto
import java.time.LocalDate
import java.time.ZoneOffset

/**
 * The in-room check-in validation chain — a faithful port of
 * apps/coordinator-pwa/src/qr/validator.ts (and the Go gateway gate order).
 * Pure: all I/O goes through [Store] and the injected clock/id functions, so the
 * whole chain is unit-tested off-device. Any failure returns immediately.
 */
class CheckinValidator(
    private val store: Store,
    private val nowMillis: () -> Long = { System.currentTimeMillis() },
    private val nowIso: () -> String = { java.time.Instant.ofEpochMilli(System.currentTimeMillis()).toString() },
    private val newUuid: () -> String = { java.util.UUID.randomUUID().toString() },
) {
    fun validate(rawQr: String, session: ActiveSession, device: DeviceContext): ValidationResult {
        // ── Step 1: parse + RSA-2048 signature ──────────────────────────────────
        val qr = FlatJson.parseQr(rawQr)
            ?: return ValidationResult(ValidationStatus.REJECTED, RejectionReason.INVALID_SIGNATURE)
        val f = qr.fields
        val body = QrVerify.canonicalBody(
            QrVerify.QrPayload(f.studentId, f.tenantId, f.courseId, f.fullName, f.academicYear, f.serialNumber, f.expiryDate, f.issuedAt)
        )
        val sigOk = try { QrVerify.verify(session.institutionPublicKeyPem, body, qr.signatureB64) } catch (_: Exception) { false }
        if (!sigOk) return ValidationResult(ValidationStatus.REJECTED, RejectionReason.INVALID_SIGNATURE)

        // ── Step 2: expiry (malformed date ⇒ expired) ───────────────────────────
        val expiryMs = try {
            LocalDate.parse(f.expiryDate).atStartOfDay(ZoneOffset.UTC).toInstant().toEpochMilli()
        } catch (_: Exception) { return ValidationResult(ValidationStatus.REJECTED, RejectionReason.QR_EXPIRED) }
        if (expiryMs < nowMillis()) return ValidationResult(ValidationStatus.REJECTED, RejectionReason.QR_EXPIRED)

        // ── Step 3: tenant match ─────────────────────────────────────────────────
        if (f.tenantId != session.tenantId)
            return ValidationResult(ValidationStatus.REJECTED, RejectionReason.TENANT_MISMATCH)

        val studentIdHash = VaultCrypto.hmacHex(session.studentHashKey, f.studentId)

        // ── Step 6a: cross-student device lock ───────────────────────────────────
        store.bindingByFingerprint(device.fingerprintHash)?.let {
            if (it.studentIdHash != studentIdHash)
                return ValidationResult(ValidationStatus.REJECTED, RejectionReason.DEVICE_BELONGS_TO_ANOTHER_STUDENT)
        }

        // ── Step 4: roster lookup ────────────────────────────────────────────────
        if (studentIdHash !in session.rosterHashes)
            return ValidationResult(ValidationStatus.REJECTED, RejectionReason.NOT_ON_ROSTER)

        // ── Step 4b: serial must match the current issued serial ─────────────────
        val currentSerial = session.rosterSerials[studentIdHash]
        if (currentSerial == null || f.serialNumber != currentSerial)
            return ValidationResult(ValidationStatus.REJECTED, RejectionReason.SERIAL_REVOKED)

        // ── Step 5: hardware fingerprint (bind on first scan) ────────────────────
        val stored = store.bindingByStudent(studentIdHash)
        if (stored != null) {
            if (stored.fingerprintHash != device.fingerprintHash)
                return ValidationResult(ValidationStatus.REJECTED, RejectionReason.DEVICE_MISMATCH, auditFlag = "DEVICE_MISMATCH")
        } else {
            store.putBinding(DeviceBinding(studentIdHash, device.fingerprintHash, session.academicYear, nowIso()))
        }

        // ── Step 6: duplicate scan ───────────────────────────────────────────────
        if (store.hasAttendance(session.sessionId, studentIdHash))
            return ValidationResult(ValidationStatus.REJECTED, RejectionReason.DUPLICATE_SCAN)

        // ── Step 7: one device per session ───────────────────────────────────────
        if (store.deviceUsedByOther(session.sessionId, device.fingerprintHash, studentIdHash))
            return ValidationResult(ValidationStatus.REJECTED, RejectionReason.DEVICE_ALREADY_USED)

        // ── PRESENT: append-only record ──────────────────────────────────────────
        store.addAttendance(
            AttendanceRecord(
                logId = newUuid(),
                sessionId = session.sessionId,
                studentIdHash = studentIdHash,
                deviceFingerprintHash = device.fingerprintHash,
                sequenceNumber = store.attendanceCount(session.sessionId) + 1,
                checkinTimestamp = nowIso(),
            )
        )
        return ValidationResult(ValidationStatus.PRESENT, studentIdHash = studentIdHash)
    }
}

File: frontend/coordinator-flutter/lib/engine/checkin_validator.dart
import 'dart:math';

import '../crypto/qr_verify.dart' show QrVerify, QrPayload;
import '../crypto/vault_crypto.dart' show VaultCrypto;
import 'flat_json.dart' show FlatJson;
import 'models.dart';
import 'store.dart' show Store;

class CheckinValidator {
  final Store store;
  final bool enforceDeviceLock;
  final int Function() nowMillis;
  final String Function() nowIso;
  final String Function() newUuid;

  CheckinValidator(
    this.store, {
    this.enforceDeviceLock = true,
    int Function()? nowMillis,
    String Function()? nowIso,
    String Function()? newUuid,
  })  : nowMillis = nowMillis ?? _defaultNowMillis,
        nowIso = nowIso ?? _defaultNowIso,
        newUuid = newUuid ?? _defaultNewUuid;

  static int _defaultNowMillis() => DateTime.now().millisecondsSinceEpoch;

  static String _defaultNowIso() => DateTime.now().toUtc().toIso8601String();

  static String _defaultNewUuid() {
    final rng = Random.secure();
    final b = List<int>.generate(16, (_) => rng.nextInt(256));
    b[6] = (b[6] & 0x0f) | 0x40;
    b[8] = (b[8] & 0x3f) | 0x80;
    final h = b.map((x) => x.toRadixString(16).padLeft(2, '0')).join();
    return '${h.substring(0, 8)}-${h.substring(8, 12)}-${h.substring(12, 16)}-'
        '${h.substring(16, 20)}-${h.substring(20)}';
  }

  ValidationResult validate(String rawQr, ActiveSession session, DeviceContext device) {
    final qr = FlatJson.parseQr(rawQr);
    if (qr == null) {
      return const ValidationResult.rejected(RejectionReason.invalidSignature);
    }
    final f = qr.fields;
    final body = QrVerify.canonicalBody(QrPayload(
      f.studentId,
      f.tenantId,
      f.courseId,
      f.fullName,
      f.academicYear,
      f.serialNumber,
      f.expiryDate,
      f.issuedAt,
    ));
    final sigOk = _try(() => QrVerify.verify(session.institutionPublicKeyPem, body, qr.signatureB64));
    if (!sigOk) {
      return const ValidationResult.rejected(RejectionReason.invalidSignature);
    }

    final expiryMs = _parseExpiry(f.expiryDate);
    if (expiryMs == null || expiryMs < nowMillis()) {
      return const ValidationResult.rejected(RejectionReason.qrExpired);
    }

    if (f.tenantId != session.tenantId) {
      return const ValidationResult.rejected(RejectionReason.tenantMismatch);
    }

    final studentIdHash = VaultCrypto.hmacHex(session.studentHashKey, f.studentId);

    final lock = _rosterAndDeviceLock(studentIdHash, session, device);
    if (lock != null) {
      return ValidationResult.rejected(lock);
    }

    final currentSerial = session.rosterSerials[studentIdHash];
    if (currentSerial == null || f.serialNumber != currentSerial) {
      return const ValidationResult.rejected(RejectionReason.serialRevoked);
    }

    return _bindAndRecord(studentIdHash, session, device);
  }

  ValidationResult validateReg(String regNumber, ActiveSession session, DeviceContext device) {
    final reg = regNumber.trim();
    if (reg.isEmpty) {
      return const ValidationResult.rejected(RejectionReason.notOnRoster);
    }
    final studentIdHash = VaultCrypto.hmacHex(session.studentHashKey, reg);
    final lock = _rosterAndDeviceLock(studentIdHash, session, device);
    if (lock != null) {
      return ValidationResult.rejected(lock);
    }
    return _bindAndRecord(studentIdHash, session, device);
  }

  RejectionReason? _rosterAndDeviceLock(String studentIdHash, ActiveSession session, DeviceContext device) {
    if (enforceDeviceLock) {
      final binding = store.bindingByFingerprint(device.fingerprintHash);
      if (binding != null && binding.studentIdHash != studentIdHash) {
        return RejectionReason.deviceBelongsToAnotherStudent;
      }
    }
    if (!session.rosterHashes.contains(studentIdHash)) {
      return RejectionReason.notOnRoster;
    }
    return null;
  }

  ValidationResult _bindAndRecord(String studentIdHash, ActiveSession session, DeviceContext device) {
    final stored = store.bindingByStudent(studentIdHash);
    if (stored != null) {
      if (enforceDeviceLock && stored.fingerprintHash != device.fingerprintHash) {
        return const ValidationResult.rejected(
          RejectionReason.deviceMismatch,
          auditFlag: 'DEVICE_MISMATCH',
        );
      }
    } else {
      store.putBinding(DeviceBinding(
        studentIdHash,
        device.fingerprintHash,
        session.academicYear,
        nowIso(),
      ));
    }
    if (store.hasAttendance(session.sessionId, studentIdHash)) {
      return const ValidationResult.rejected(RejectionReason.duplicateScan);
    }
    if (enforceDeviceLock &&
        store.deviceUsedByOther(session.sessionId, device.fingerprintHash, studentIdHash)) {
      return const ValidationResult.rejected(RejectionReason.deviceAlreadyUsed);
    }
    store.addAttendance(AttendanceRecord(
      logId: newUuid(),
      sessionId: session.sessionId,
      studentIdHash: studentIdHash,
      deviceFingerprintHash: device.fingerprintHash,
      sequenceNumber: store.attendanceCount(session.sessionId) + 1,
      checkinTimestamp: nowIso(),
    ));
    return ValidationResult.present(studentIdHash: studentIdHash);
  }

  static int? _parseExpiry(String date) {
    final dt = DateTime.tryParse(date);
    if (dt == null) return null;
    return DateTime.utc(dt.year, dt.month, dt.day).millisecondsSinceEpoch;
  }

  static bool _try(bool Function() f) {
    try {
      return f();
    } catch (_) {
      return false;
    }
  }
}